Data-Structures/Linked-List/CycleDetectionII.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * A LinkedList based solution for Detecting a start node of the Cycle in a list.
+ * https://en.wikipedia.org/wiki/Cycle_detection
+ */
+
+function detectCycleNode(head) {
+  /*
+  Problem Statement:
+  Given the head of a linked list, return the node where the cycle begins. If there is no cycle, return null.
+  Link for the Problem: https://leetcode.com/problems/linked-list-cycle-ii/
+  */
+
+  let length = 0
+  let fast = head
+  let slow = head
+
+  while (fast !== null && fast.next !== null) {
+    fast = fast.next.next
+    slow = slow.next
+    if (fast === slow) {
+      // Calculate length of the cycle
+      length = lengthCycle(slow)
+      break
+    }
+  }
+
+  if (length === 0) {
+    // If length = 0, return null
+    return null
+  }
+
+  //Point the nodes at head
+  fast = head
+  slow = head
+  // Move slow pointer ahead by 'length' of cycle times
+  while (length > 0) {
+    slow = slow.next
+    length--
+  }
+
+  // Now move both pointers until they meet - this will be the start of cycle
+  while (fast !== slow) {
+    fast = fast.next
```

Comment:
This would have to be `fast.next.next`. If this is not failing the tests in its current form, it means `fast == slow`, and all this code is practically dead.

Reply:
I appreciate you pointing out the potential for misinterpretation. Here, the `fast` pointer is kept stationery at the head only and `slow` is moved ahead first in the list. And then we move both one by one to make them meet at the starting node. Please see the attached images above for my justification.

Reply:
Yeah, I see why now. Again, see my point about the confusing naming of these variables.